src/OpenTK.GraphicsLibraryFramework/GLFW.cs

```diff
@@ -143,6 +143,62 @@ protected GLFW(string path, ImplementationOptions options)
         /// <inheritdoc />
         public abstract unsafe float* GetJoystickAxes(int jid, out int count);
 
+        /// <inheritdoc />
+        public abstract unsafe byte* GetJoystickButtons(int jid, out int count);
+
+        /// <inheritdoc />
+        public abstract unsafe byte* GetJoystickHats(int jid, out int count);
+        
+        /// <inheritdoc />
+        public abstract string GetJoystickName(int jid);
+        
+        /// <inheritdoc />
+        public abstract string GetJoystickGUID(int jid);
+
+        /// <inheritdoc />
+        public abstract void SetJoystickUserPointer(int jid, IntPtr ptr);
+
+        /// <inheritdoc />
+        public abstract IntPtr GetJoystickUserPointer(int jid);
+
+        /// <inheritdoc />
+        public abstract bool JoystickIsGamepad(int jid);
+
+        /// <inheritdoc />
+        public abstract bool UpdateGamepadMappings(string newMapping);
+
+        /// <inheritdoc />
+        public abstract string GetJoystickname(int jid);
+        
+        /// <inheritdoc />
+        public abstract bool GetGamepadState(int jid, out GamepadState state);
+
+        /// <inheritdoc />
+        public abstract double GetTime();
+
+        /// <inheritdoc />
+        public abstract void SetTime(double time);
+        
+        /// <inheritdoc />
+        public abstract long GetTimerValue();
+        
+        /// <inheritdoc />
+        public abstract long GetTimerFrequency();
+        
+        /// <inheritdoc />
+        public abstract unsafe Window* GetCurrentContext();
+        
+        /// <inheritdoc />
+        public abstract unsafe void SwapBuffers(Window* window);
+        
+        /// <inheritdoc />
+        public abstract bool ExtensionSupported(string extensionName);
+
+        /// <inheritdoc />
+        public abstract bool VulkanSupported();
+
+        public abstract unsafe char** GetRequiredInstanceExtensions(out int count);
```

Comment:
Double depth character pointers can be mapped to `string[]`

https://github.com/opentk/opentk/blob/master/src/Generator.Bind/Specifications/csharp.tm#L25

Reply:
There were already discussions on some parts and as far as I understood we want to have low level versions of wrappers available and the nicer to use functions optionally. As the mapping to string array isn't always someone wants(e.g. not on heap or whatever)...

Reply:
Introducing this map will make the GLFW bindings uniform to our OpenGL and OpenCL bindings. It will also be uniform to the rest of the most recent commit, seeing as this same commit changed all character pointers into strings.

Reply:
Afaik there currently aren't any OpenGL bindings(but a Generator which isn't even complete yet as well)? As for OpenCL we shouldn't orient us at something which was never close to complete imho.
Apart from that I would be highly against a strict usage of string instead of char*, instead I would like to have both. Because strings have a pretty high performance load on native calls. Apart from often a lot of micro allocations, which can often be prevented.

Reply:
There are OpenGL bindings but they're in a work-in-progress alongside the binder. Both the new and old binder produce string instead of char* as it's part of the typemap. If you're against this I will have to direct you to @Nihlus

Reply:
I remember asking about converting char** to string, I thought Nihlus said ADL couldn't do that due to a lack of information on the array length.

Reply:
Ah yes.
![image](https://user-images.githubusercontent.com/11160611/53698367-a957ad00-3dd3-11e9-86c8-d35617dab142.png)

Will need to change this in the binder.

Reply:
Yes, you cannot return arrays from native functions, as it ends up being just a pointer. There's no way of knowing the number of elements returned from the array, and as such no way of knowing the length of the pointer.

Array returns should be marshalled as pointers, and converted in the mixed-mode class to managed arrays using the sibling length function that typically accompany array-returning methods. e.g, `int* GetMyValues` and `size_t GetNumMyValues`.

Reply:
@Perksey Something to keep in mind is that `passing` arrays works fine. It's when you're returning them that issues crop up.

Reply:
Noted, thanks.

Reply:
This can be closed.